### PR TITLE
First stab at ancestry resolution text

### PR DESCRIPTION
**Pull request by jeromekelleher**
WIP

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Here's a stab at some formal definitions (brought togther in one place for clarity here):

formal-defs

Any thoughts here @JereKoskela @hyanwong? I'm trying hard for mathematical economy here, no necessarily mapping to computational implementations.

**Comment by jeromekelleher** [comment]
Obviously would have to define what we mean by "through the local trees" here.

**Comment by hyanwong** [comment]
Rather than using N (natural numbers), do we want to define N to be the set of all nodes? Then it's a bt more obvious to say S is a subset of the set of all nodes.

Also, there is a constraint on the eARG that says that there only two parents per child. In the original formulation, I suspect there is also the constraint that there are only 2 children per parent, although that's not strictly necessary.

**Comment by hyanwong** [comment]
> Obviously would have to define what we mean by "through the local trees" here.

Do we need to say "a path *down* from c" (and do we need to indicate that this is a DAG, i.e. that the order of (c, p) matters)

> The "non-empty subset" is also intended to imply that c is not a root

Are we not also allowing nodes above the root? I think that';s OK, right? Just not what we normally bother doing.

**Comment by hyanwong** [comment]
Presumably we also define what an implicit vs an explicit gARG is?

**Comment by jeromekelleher** [comment]
> But rather than using N (natural numbers), do we want to define N to be the set of all nodes? Then it's a bt more obvious to say S is a subset of the set of all nodes.

I don't think we need to define the set of nodes, it's implicitly defined by whether a number exists as a child or parent in E. We probably don't need to say that S is a subset of the natural numbers though, and can just say "S is a set of nodes".

Good point on the restrictions on the EARG, they need to be stated.

**Comment by jeromekelleher** [comment]
> Presumably we also define what an implicit vs an explicit gARG is?

That's the distinction between a GARG and a sample resolved GARG. You can express a GARG without defining any samples. Then all you have is the local parent-child relationships. If you define a set of samples, you can then resolve the ancestry wrt to that set.

**Comment by jeromekelleher** [comment]
> Are we not also allowing nodes above the root(s)? I think that's OK, right? Just not what we normally bother doing.

Ah, you're right, that's full simplification I'm thinking about. No need for that then.

**Comment by jeromekelleher** [comment]
> It's the same tuple as in the GARG case, subject to a further constraint, right? In which case we could economise a bit and just say that.

Yes, strictly it's the same exception the interval I_{resolved} will be \subseteq I_{not resolved}.

**Comment by hyanwong** [comment]
> > Presumably we also define what an implicit vs an explicit gARG is?
> 
> That's the distinction between a GARG and a sample resolved GARG. You can express a GARG without defining any samples. Then all you have is the local parent-child relationships. If you define a set of samples, you can then resolve the ancestry wrt to that set.

But you can have a sample-resolved gARG which defines the edges as (0, L] or (0,x] or (x, L), and you can also have a sample-resolved gARG which defines them so that there are no hanging edges in the local trees. I think both are valid sample resolved gARGs, from your definition above, unless I've missed something.

**Comment by jeromekelleher** [comment]
> But you can have an implicit sample-resolved gARG which defines the edges as (0, L] or (0,x] or (x, L), and you can also have an explicit sample-resolved gARG which defines them so that there are no hanging edges in the local trees. I think both are valid sample resolved gARGs, from your definition above, unless I've missed something.

No, you can only define "hanging edges" in the trees if you define some set of samples. If you don't define a set of samples, then every node in the graph must be consider a sample, and therefore (by definition) you can't have any hanging edges.

There's no implicit-vs-explicit distinction here, the important thing is whether you define some of the nodes as samples or not (and then percolate the ancestry). You could have "samples-but-not-resolved" as well, but what's the point?

**Comment by jeromekelleher** [comment]
> True, but sigma: N -> N_0 only needs to be defined for N (the set of all nodes) rather than N (all the natural numbers)

Like I said, this isn't about practical implementation. It's mathematically more economical to avoid having to define what the set of nodes is.

@JereKoskela - are you comfortable with this or am I on my own here?

**Comment by hyanwong** [comment]
> No, you can only define "hanging edges" in the trees if you define some set of samples. If you don't define a set of samples, then every node in the graph must be consider a sample, and therefore (by definition) you can't have any hanging edges.

I think I misunderstood what you meant by "through the local trees".

> There's no implicit-vs-explicit distinction here, the important thing is whether you define some of the nodes as samples or not (and then percolate the ancestry). You could have "samples-but-not-resolved" as well, but what's the point?

Yes, I thought you meant "samples-but-not-resolved", sorry. 

P.s. did @JereKoskela  say that there was a standard symbol we could return for "no breakpoint", rather than 0?

**Comment by hyanwong** [comment]
By "set of disjoint intervals over which c inherits from p", I thought you meant *locally* (implicit), rather than w.r.t. the samples. Perhaps that needs more clarification?

Does a (non-sample-resolved) gARG require those constraints (that the edges above a recombination node sum up to fill the entire interval)? I think maybe... not?

**Comment by JereKoskela** [comment]
> Like I said, this isn't about practical implementation. It's mathematically more economical to avoid having to define what the set of nodes is.

I don't see a problem with sigma being defined on a domain which is bigger than what we actually need. As long as it's well-defined for each argument we actually want to plug in, it's good in my book.

> a standard symbol we could return for "no breakpoint", rather than 0?

\Delta or \partial are common notations for so-called cemetery states for transient stochastic processes. We could try to co-opt one (probably \Delta) for our purposes as a kind of null state. I'd prefer a symbol which has no independent meaning so that it's clearly arbitrary and we don't intend to do anything with it, but I'm pretty indifferent about what that symbol should be. 

> Then the codomain of sigma becomes more complicated. This is more elegant because a breakpoint at 0 means the interval to the left is empty.

On the other hand, this sounds like a good reason for a symbol with some independent meaning.

**Comment by hyanwong** [comment]
> Then the codomain of sigma becomes more complicated. This is more elegant because a breakpoint at 0 means the interval to the left is empty.

OK, I guess we need to say that X is a positive real then?

**Comment by hyanwong** [comment]
Off the back of https://github.com/tskit-dev/what-is-an-arg-paper/issues/56#issuecomment-1040717922, do we not (normally) define the eARG and the gARG with respect to a set of samples, such that the edges extend backwards from the samples? Perhaps that's not necessary, but I think it is why the structure is called the *ancestral* recombination graph, rather than just the "recombination graph", and it's certainly in the spirit of the original formulations.

The more general structure where *all* lines of ancestry are traced, both forwards and backwards in time, is perhaps the "genetic pedigree", or simply the "recombination graph"?

There's a difference between a "sample-based gARG" and a "sample-*resolved* gARG", though. It's the resolution stuff that implies running the Hudson/simplification algorithm.

**Comment by jeromekelleher** [comment]
This confusion is probably my fault - I'm really talking literally about the encoding. The EARG  and GARG encodings are local and don't refer to a set of samples. So, you could, say, add a new sample at the leaves, or decide that one of the existing internal nodes was a sample, and this wouldn't alter any of the rest of the graph. The "sample resolved GARG" is not local: it's always wrt to some set of samples, and changing the samples will effect other parts of the graph.

So it's not about the model or the process or anything, just literally about the lists of tuples that are used to encode the structure.

File: argutils/ancestry.py [hunk 1]
```diff
@@ -572,3 +573,127 @@ def ca(*args):
...
+            print("parent = ", p)
+            print("Ip = ", I[p])
+            print("Ic = ", I[c])
+            print("e = ", e[2])
+            print("intersection = ", I[c].intersection(e[2]))
+            I[p] = I[p].union(I[c].intersection(e[2]))
+            N.add(p)
+    return [(c, p, I[c]) for c, p, _ in E]
+
+def as_earg(ts):
```
Comment (jeromekelleher) [inline argutils/ancestry.py:637]: These two functions might help clarify things @hyanwong - this is a literal implementation of the EARG definition and a reasonable method for recovering a local tree from it.

(I'm still working on the GARG and resolved GARG versions)

File: argutils/ancestry.py [hunk 2]
```diff
@@ -572,3 +573,127 @@ def ca(*args):
...
+    Given a minimal EARG definition (E, sigma), return the tree for a the given
+    set of samples at the specified position as a dictionary parent->child.
+    """
+    N = set(S)
+    parent = {}
+    while len(N) > 0:
+        c = N.pop()
+        P = [e[1] for e in E if e[0] == c]
+        if len(P) > 0:
+            if x < sigma[c]:
```
Comment (jeromekelleher) [inline argutils/ancestry.py:675]: Note: this is why I like defining sigma(u) = inf for non recombination nodes. This is about as cleanly as you can implement this operation.

**Comment by hyanwong** [comment]
> The EARG and GARG encodings are local and don't refer to a set of samples. So, you could, say, add a new sample at the leaves, or decide that one of the existing internal nodes was a sample, and this wouldn't alter any of the rest of the graph. 

Yes, ISWYM. I hadn't noticed the lack of the reverse-time approach in the original definitions. I just thought it worth pointing it out that the encoding that you define will be appropriate for any genome pedigree graph, not just the (G/E)ARG.

> The "sample resolved GARG" is not local: it's always wrt to some set of samples, and changing the samples will affect other parts of the graph.

That's a very good point.

File: argutils/ancestry.py [hunk 3]
```diff
@@ -572,3 +573,127 @@ def ca(*args):
...
+        edge = next(edges, None)
+    # We could wrap sigma as a function to be literal about the definition
+    # but this is simpler for debugging.
+    return E, sigma
+
+
+def earg_get_tree(E, sigma, S, x):
+    """
+    Given a minimal EARG definition (E, sigma), return the tree for a the given
+    set of samples at the specified position as a dictionary parent->child.
```
Comment (jeromekelleher) [inline argutils/ancestry.py:667]: Note to self - comments are wrong dictionary is mapping child->parent

File: argutils/ancestry.py (hunk 2 above)
Comment (hyanwong) [reply argutils/ancestry.py:675]: I can see why, computationally, you might want to use `inf`, but that's not the same as using it in the mathematical definition, is it?

Comment (JereKoskela) [reply argutils/ancestry.py:675]: Yan beat me to it, but I was going to make the same point. I'm not strongly opposed to using inf in the text, but it always stands out to me as a bit odd when looking through the draft.

Comment (jeromekelleher) [reply argutils/ancestry.py:675]: Let's see how it goes - I have some draft text describing how we build the trees in the text, and it does help make that description simpler and more concise. We may or may not want to include this. Maybe it's *too* neat though! Anyway, that's my motivation - let's leave it for now.

**Comment by hyanwong** [comment]
A minor point, but given the comments above about the mathematical description specifying the encoding rather than the topology of the ARG , I wonder if instead of saying "we define an eARG as ...", we should say "we encode an eARG as..."? Personally, I think of the *definition* as also including restrictions on the topology, whereas the encoding does not have to.

**Comment by jeromekelleher** [comment]
All of the requirements should be present in the definition - if they're not, it's not a full definition. An encoding is just a way of being mathematically precise about what we're talking about. The encoding absolutely does define the topology - that's the whole point.

**Comment by hyanwong** [comment]
> All of the requirements should be present in the definition - if they're not, it's not a full definition. An encoding is just a way of being mathematically precise about what we're talking about. The encoding absolutely does define the topology - that's the whole point.

A (standard Griffiths) ARG defines sample nodes (or more precisely sampling events), and the topology is defined above those. That's not captured in your definition of an eARG?

**Comment by hyanwong** [comment]
We also need to specify that this is directed (perhaps a given, since we talk about parents and child nodes), and that we can't have cycles, right. We could just say it is a type of DAG.

**Comment by jeromekelleher** [comment]
> A (standard Griffiths) ARG defines sample nodes (or more precisely sampling events), and the topology is defined above those.

Have you got a reference for this? I've struggled to find anything precise.

**Comment by hyanwong** [comment]
> > A (standard Griffiths) ARG defines sample nodes (or more precisely sampling events), and the topology is defined above those.
> 
> Have you got a reference for this? I've struggled to find anything precise.

Griffiths & Majoram define the ARG as the ancestry, backwards in time, of a sample of `n` genes at time t=0. So that's equivalent, right?

**Comment by JereKoskela** [comment]
> We also need to specify that this is directed (perhaps a given, since we talk about parents and child nodes), and that we can't have cycles, right. We could just say it is a type of DAG.

I don't know whether this is usually present in ARG definitions, but the coming down from infinity property of the Kingman coalescent implies that it is possible to construct an ARG for a whole, countably infinite population. I guess you could view that as sampling everyone, but I think that's only a semantic difference from modelling a whole population without sampling as such.

**Comment by jeromekelleher** [comment]
> Griffiths & Majoram define the ARG as the ancestry, backwards in time, of a sample of n genes at time t=0. So that's equivalent, right?

No, I'm talking about the *structure* here. 

Can we leave this discussion until I have some draft text that's ready for review please?

**Comment by hyanwong** [comment]
> > A (standard Griffiths) ARG defines sample nodes (or more precisely sampling events), and the topology is defined above those. That's not captured in your definition of an eARG?
> 
> I don't know whether this is usually present in ARG definitions, but the coming down from infinity property of the Kingman coalescent implies that it is possible to construct an ARG for a whole, countably infinite population. I guess you could view that as sampling everyone, but I think that's only a semantic difference from modelling a whole population without sampling as such.

Yes, by "sample", it could be the entire population. But actually it's not the "sampling" itself that I'm thinking about, it's the fact that the ARG is defined as the ancestry backwards in time from a set of specified nodes. I imagine that you could build a recombination graph that includes nodes that are not ancestral to the specified nodes. That would not, I think, be an ARG?

**Comment by hyanwong** [comment]
> Can we leave this discussion until I have some draft text that's ready for review please?

Yep - sorry for the noise. Let's just wait. Just copying one of my comments above in case I forget too:

> Also, there is a constraint on the eARG that says that there only two parents per child

**Comment by jeromekelleher** [comment]
I've done a bunch of stuff here, and I'm no closer to fixing the notation etc, so it's probably best to just merge this much. Definitely making forward progress though.

**Comment by hyanwong** [comment]
This looks good, and I think we should merge, then change.

My thinking on event ARGs vs general event-based recombination graphs has clarified a little here, and perhaps it would be better continuing the discussion in a specific issue after merging?

**Comment by jeromekelleher** [comment]
> My thinking on event ARGs vs general event-based recombination graphs has clarified a little here, and perhaps it would be better continuing the discussion in a specific issue after merging?

Sure

**Pull request merged**